### PR TITLE
[Feature]: Support ImageNet-A,R,S,C

### PR DESCRIPTION
**Pull request by YuanLiuuuuuu**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation
Add evaluation on ImageNet-A,R,S,C

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmcls/evaluation/metrics/single_label.py [hunk 1]
```diff
@@ -87,6 +87,9 @@ class Accuracy(BaseMetric):
             names to disambiguate homonymous metrics of different evaluators.
             If prefix is not provided in the argument, self.default_prefix
             will be used instead. Defaults to None.
+        ano_file_path (str, optional): The path of the annotation file. This
```
Comment (mzr1996) [inline mmcls/evaluation/metrics/single_label.py:90]: `ann_file` to align with dataset classes.

File: mmcls/evaluation/metrics/single_label.py [hunk 2]
```diff
@@ -136,6 +140,18 @@ def __init__(self,
...
+        if ann_file_path is not None:
+            with open(ann_file_path, 'r') as f:
+                labels = [
+                    int(item.strip().split()[-1]) for item in f.readlines()
+                ]
+            label_dict = {label: 1 for label in labels}
+            self.label_mask = torch.tensor(
+                [1 if i in label_dict else 0 for i in range(1000)]).cpu()
+        else:
+            self.label_mask = torch.tensor([1] * 1000).cpu()
```
Comment (mzr1996) [inline mmcls/evaluation/metrics/single_label.py:153]: The hard-coded 1000 is a little wried because it's only for ImageNet variants. Better to save available categories indices.

File: mmcls/evaluation/metrics/single_label.py [hunk 3]
```diff
@@ -151,8 +167,10 @@ def process(self, data_batch, data_samples: Sequence[dict]):
             result = dict()
             pred_label = data_sample['pred_label']
             gt_label = data_sample['gt_label']
+            result['img_path'] = data_sample['img_path']
```
Comment (mzr1996) [inline mmcls/evaluation/metrics/single_label.py:170]: If the `img_path` is used in the `mCE` metric, please add it to the result in `mCE`, because `Accuracy` doesn't need it.

File: mmcls/evaluation/metrics/mce.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    return ce_alexnet
+
+
+@METRICS.register_module()
+class mCE(Accuracy):
+    """Mean Corruption Error (mCE) metric.
+
+    The mCE metric is proposed in `Benchmarking Neural Network Robustness to
+    Common Corruptions and Perturbations
+    <https://openreview.net/pdf?id=HJz6tiCqYm>`_.
```
Comment (mzr1996) [inline mmcls/evaluation/metrics/mce.py:38]: Maybe replace it with the Arxiv link https://arxiv.org/abs/1903.12261

File: mmcls/evaluation/metrics/mce.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    ce_alexnet['contrast'] = 0.853204
+    ce_alexnet['elastic_transform'] = 0.646056
+    ce_alexnet['pixelate'] = 0.717840
+    ce_alexnet['jpeg_compression'] = 0.606500
+
+    return ce_alexnet
+
+
+@METRICS.register_module()
+class mCE(Accuracy):
```
Comment (mzr1996) [inline mmcls/evaluation/metrics/mce.py:33]: Better to use `CorruptionError` since `CE` is often used as the abbreviation of Cross Entropy. And remove the `mean` in the class name so that we can expand the class to output in different modes.

File: docs/en/user_guides/finetune.md [hunk 6]
```diff
@@ -228,3 +228,61 @@ It's because our training schedule is for a batch size of 128. If using 8 GPUs,
...
+
+## Evaluate the fine-tuned model on ImageNet variants
+
+It's a common practice to evaluate the ImageNet-(1K, 21K) fine-tuned model the ImageNet-1K validation set. This set
+shares similar data distribution with the training set, but in real world, the inference data is more likely to share
+different data distribution with the training set. To have a full evaluation of model's performance on
+out-of-distribution datasets, research community introduces the ImageNet-variant datasets, which shares different data
+distribution with that of ImageNet-(1K, 21K)., MMClassification supports evaluating the fine-tuned model on
+[ImageNet-A](https://arxiv.org/pdf/1907.07174v4.pdf), [ImageNet-R](https://arxiv.org/pdf/2006.16241v3.pdf),
+[ImageNet-C](https://arxiv.org/pdf/1903.12261v1.pdf), and [ImageNet-S](https://arxiv.org/pdf/2106.03149v3.pdf).
```
Comment (mzr1996) [inline docs/en/user_guides/finetune.md:240]: Better to use a full name since the abbreviation may conflict. And I prefer to provide the link to abstract pages.
```suggestion
[ImageNet-Adversarial (A)](https://arxiv.org/abs/1907.07174), [ImageNet-Rendition (R)](https://arxiv.org/abs/2006.16241),
[ImageNet-Corruption (C)](https://arxiv.org/abs/1903.12261), and [ImageNet-Sketch (S)](https://arxiv.org/abs/1905.13549).
```

File: projects/ood_eval/config/vit_ood-eval_toy-example.py [hunk 7]
```diff
@@ -0,0 +1,226 @@
+data_preprocessor = dict(
+    num_classes=1000,
+    mean=[123.675, 116.28, 103.53],
+    std=[58.395, 57.12, 57.375],
+    to_rgb=True)
+bgr_mean = [103.53, 116.28, 123.675]
+bgr_std = [57.375, 57.12, 58.395]
+train_pipeline = [
+    dict(type='LoadImageFromFile', file_client_args=dict(backend='disk')),
+    dict(
```
Comment (mzr1996) [inline projects/ood_eval/config/vit_ood-eval_toy-example.py:10]: Please clean up the config file and remove the duplicated settings.

File: tests/test_evaluation/test_metrics/test_mce.py [hunk 8]
```diff
@@ -0,0 +1,19 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from unittest import TestCase
+
+import torch
+
+from mmcls.registry import METRICS
+
+
+class TestAccuracy(TestCase):
```
Comment (mzr1996) [inline tests/test_evaluation/test_metrics/test_mce.py:9]: Change to test class name.

**Review by mzr1996 — CHANGES_REQUESTED** [review]

File: docs/en/user_guides/finetune.md [hunk 9]
```diff
@@ -228,3 +228,61 @@ It's because our training schedule is for a batch size of 128. If using 8 GPUs,
...
+test_dataloader=dict(dataset=dict(data_root='data/imagenet-x'))
+test_evaluator=dict(meta_file_path='data/imagenet-x/meta/val.txt')
+```
+
+### Start test
+
+This step is the common test step, you can follow this [guide](https://mmclassification.readthedocs.io/en/1.x/user_guides/train_test.html)
+to evaluate your fine-tuned model on out-of-distribution datasets.
+
+To make it easier, we also provide an off-the-shelf config files, for [ImageNet-C](https://github.com/open-mmlab/mmclassification/tree/dev-1.x/projects/example_project/ood_eval/vit_ood-eval_toy-example.py) and [ImageNet-C](https://github.com/open-mmlab/mmclassification/tree/dev-1.x/projects/example_project/ood_eval/vit_ood-eval_toy-example_imagnet-c.py), and you can have a try.
```
Comment (Ezra-Yu) [inline docs/en/user_guides/finetune.md:288]: The configuration file given by the project is a bit redundant. Examples can be given like this, where the original cfg can be left unchanged. Can be written directly in the doc, no need to give actual examples

```
_base_ = 'mmcls::resnet/resnetv1c50_8xb32_in1k.py'  # can be your own config

test_dataloader=dict(dataset=dict(data_root='data/imagenet-x'))
test_evaluator=dict(meta_file_path='data/imagenet-x/meta/val.txt')
```

Comment (YuanLiuuuuuu) [reply docs/en/user_guides/finetune.md:288]: It's better to keep these configuration files, such that users can try ood evaluation without any efforts. For these contents in these files, I will rewrite them according to your suggestions.

**Review by mzr1996 — APPROVED** [review]
LGTM

**Comment by LALBJ** [comment]
I saw this requirement during the OpenMMLab Code Camp event, and I would like very much to participate in solving this PR. However, based solely on the description provided during the event, I am not very clear on how to get started. Could you please advise me on how to complete this task?
